File: C Programming/Sorting/QuickSort.c
#include <stdio.h>

int partition(int a[], int low, int high) {
    int i = low, j = high, pivot = low, temp;
    while (i < j) { // checking if i and j overlapped or not
        while (a[i] <= a[pivot])// checks for bigger elements than pivot
            i++;
        while (a[j] > a[pivot])// checks for smaller elements than pivot
            j--;
        if (i < j) { // swapping a[i] and a[j]
            temp = a[i];
            a[i] = a[j];
            a[j] = temp;
        }
    }
    // swapping a[j] and a[pivot] if i and j are overlapped
    temp = a[j];
    a[j] = a[pivot];
    a[pivot] = temp;

    return j;
}

void QuickSort(int a[], int low, int high) {
    int j;
    if (low < high) {
        j = partition(a, low, high);
        QuickSort(a, low, j - 1);
        QuickSort(a, j + 1, high);
    }
}

int main() {
    int i, a[10], n;
    printf("Enter the size of array: ");
    scanf("%d", &n);
    printf("Enter the numbers: ");
    for (i = 0; i < n; i++) {
        scanf("%d", &a[i]);
    }
    int low = 0;
    int high = n - 1;
    QuickSort(a, low, high);
    printf("Sorted array is: ");
    for (i = 0; i < n; i++) {
        printf("%d ", a[i]);
    }
    return 0;
}

File: Java/Code/QuickSort.java
import java.util.Scanner;

public class QuickSort {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the number of names: ");
        int n = sc.nextInt();
        sc.nextLine(); // Consume newline character
        System.out.println("Enter the names:");
        String[] names = new String[n];
        for (int i = 0; i < n; i++) {
            names[i] = sc.nextLine();
        }
        int low = 0;
        int high = n - 1;
        quickSort(names, low, high);
        System.out.println("\nSorted names:");
        for (String name : names) {
            System.out.println(name);
        }
        sc.close();
    }

    private static int partition(String[] names, int low, int high) {
        int i = low, j = high;
        String pivot = names[low];
        while (i < j) {
            while (/*i < j &&*/ names[i].compareTo(pivot) <= 0) // checks for bigger or equal elements than pivot
                i++;
            while (names[j].compareTo(pivot) > 0) // checks for smaller elements than pivot
                j--;
            if (i < j) { // swapping names[i] and names[j]
                String temp = names[i];
                names[i] = names[j];
                names[j] = temp;
            }
        }
        // swapping names[j] and names[low] if i and j are overlapped
        String temp = names[j];
        names[j] = names[low];
        names[low] = temp;

        return j;
    }

    private static void quickSort(String[] names, int low, int high) {
        if (low < high) {
            int partitionIndex = partition(names, low, high);
            quickSort(names, low, partitionIndex - 1);
            quickSort(names, partitionIndex + 1, high);
        }
    }
}
